Document Etherlink's transaction fees (#78)
etherlinkcom/docs: docs/network/fees.md

---
title: Fee structure
---

Etherlink transactions include two fees:

> - The _gas fee_ changes depending on Etherlink throughput over time.

Reviewer: `execution fee`?

Author: Changed to:

> Etherlink transactions include two fees:
>
>    - The execution fee, sometimes known as the gas fee, is a fee for running the transaction. It changes depending on Etherlink throughput over time; at times of high demand for Etherlink transactions, the gas fee rises.
>    - The inclusion fee helps defray the cost of the data that Etherlink stores on Tezos layer 1. Transactions that use more data pay a higher fee.

As the demand for Etherlink transactions increases, the gas fee rises.

> - The _storage fee_ is proportional to the amount of data that Etherlink stores on Tezos layer 1.

Reviewer: storage fee might be slightly misleading?  perhaps `inclusion fee` would be a better term - it's the cost of including the tx in an L1 block, as opposed to storing it permanently in the L1 context/storage

Author: Changed to "inclusion fee."

Transactions that use more data pay a higher fee.

Unlike some other chains, Etherlink does not use a voluntary gas fee, also known as a tip, to encourage block producers to include transactions more quickly.

> Because the Etherlink sequencer orders transactions in first-come-first-serve order, there is no need to offer higher fees for faster inclusion.

Author: > Etherlink ignores any fees included in the transaction beyond the execution fee and the inclusion fee.

## Gas fee

> The gas fee changes based on the transaction throughput over time.

Reviewer: the `gas price` changes

Author: > The execution fee is based on the gas price, which changes based on the transaction throughput over time.

Etherlink measures throughput in terms of _ticks_, which are processing operations.

> As demand increases, Etherlink increases the number of ticks per second to process more transactions in the same amount of time.

Reviewer: Is this technically accurate? Etherlink is fundamentally not really doing anything different here - it's more that it's requested to increase the number of execution steps/second that it processes. But the rollup node is not guaranteed to actually be able to do that. But the rollup applies back-pressure by increases the gas price.

Etherlink has a target number of ticks per second, referred to as the _speed limit_.

> To calculate the actual throughput, Etherlink stores the number of actual ticks and subtracts the speed limit every second.

Reviewer: ```suggestion
To calculate the actual throughput, Etherlink stores the number of actual ticks used and subtracts the speed limit every second.
```

Reviewer: I would maybe avoid using `actual ticks`? the tick number is an estimation by the kernel, of how many ticks it thinks have been used

Author: Changed to "the number of ticks used."

> This number of ticks that exceed the speed limit is called the _backlog_.

Reviewer: ```suggestion
The remaining number of ticks that are left over are known as the _backlog_.
```

> In this way, if the number of ticks per second exceeds the speed limit, the backlog increases and the gas fee goes up.

Reviewer: not sure if this is clear? it's less `ticks per second` and more `ticks in the current second`, if that makes sense?

If the number of ticks per second is below the speed limit, the backlog decreases or reaches zero and the gas fee decreases or reaches a base amount.
The backlog never goes below zero and the gas fee never goes below the base amount.

The gas fee depends on these parameters:

- `minimum_base_fee_per_gas`: The base fee for Etherlink transactions, which is 1 gwei
- `speed_limit`: The target number of ticks per second

> - `tolerance`: The number of ticks that the actual number of ticks per second can exceed the speed limit without increasing the gas fee

Reviewer: ```suggestion
- `tolerance`: The size the backlog is allowed to grow to, before the gas price increases
```

- `backlog`: A measure of the number of ticks per second in excess of the speed limit; Etherlink deducts the speed limit from the backlog every second
- `alpha`: A scaling factor

If the backlog is less than the tolerance, the gas fee for a transaction is `minimum_base_fee_per_gas`.

If the backlog is greater than the tolerance, Etherlink calculates the gas fee with this equation:

$$
\texttt{gas fee} = \texttt{minimum\_base\_fee\_per\_gas} * e ^{a * (\texttt{backlog} - \texttt{tolerance})}
$$

In other words, the gas fee is the base fee times Euler's number to the power of the alpha scaling factor times the backlog in excess of the tolerance.

> ## Storage fee

Reviewer: ```suggestion
## Data inclusion fee
```

Reviewer: it's an important distinction, as Etherlink may well have a notion of `storage fee` in future

The storage fee, also called the _data availability fee_, helps Etherlink cover the cost of posting data to layer 1.

The amount of the fee depends on the size of the data in the transaction (as provided by `tx.data.size()`) and the size of the access list, a list of addresses and storage keys that a transaction intends to access (as provided by `tx.access_list.size()`).

Etherlink calculates the storage fee with this equation:

> 

Author: https://github.com/etherlinkcom/docs/pull/78/commits/9e86027974e4ffb4ee32334d086331a284d0d18d

---
title: Fee structure
---

Etherlink transactions include two fees:

- The _execution fee_, sometimes known as the _gas fee_, is a fee for running the transaction.
It changes depending on Etherlink throughput over time; at times of high demand for Etherlink transactions, the gas fee rises.
This fee is burned.
- The _inclusion fee_ goes to the sequencer to defray the cost of the data that Etherlink stores on Tezos layer 1.
Transactions that use more data pay a higher fee.

Unlike some other chains, Etherlink does not use a voluntary gas fee, also known as a tip, to encourage block producers to include transactions more quickly.

> Because the Etherlink sequencer orders transactions in first-come-first-serve order, there is no need to offer higher fees for faster inclusion.

Reviewer: ```suggestion
Because the Etherlink sequencer orders transactions in first-come-first-served order, there is no need to offer higher fees for faster inclusion.
```

> Etherlink ignores any fees included in the transaction beyond the execution fee and the inclusion fee.

Reviewer: specifically, the value of `max_fee_per_gas` must be sufficient to cover this, if `max_fee_per_gas` is too low, but `max_priority_fee_per_gas` covers it, the tx will be rejected
